prompter.lisp

```diff
@@ -434,41 +483,37 @@ See `resolve-marks' for a reference on how `marks' are handled."
         (not (actions-on-current-suggestion-enabled-p source))))
 
 (export-always 'next-ready-p)
-(defun next-ready-p (prompter)
+(defun next-ready-p (prompter &key (wait-p t))
   "Block and return next PROMPTER ready source.
 It's the next source that's done updating.
 If all sources are done, return T.
 This is unblocked when the PROMPTER is `destroy'ed."
   (when prompter
-    ;; We let-bind `sync-queue' here so that it remains the same object throughout
-    ;; this function, since the slot is subject to be changed concurrently when
-    ;; the input is edited.
-    (alex:if-let ((sync-queue (sync-queue prompter)))
-      (if (= (length (ready-sources sync-queue))
-             (length (sources prompter)))
-          t
-          (calispel:fair-alt
-            ((calispel:? (ready-channel sync-queue) next-source)
-             (cond
-               ((null next-source)
-                nil)
-               (t
-                (push next-source (ready-sources sync-queue))
-                ;; Update current suggestion when update is done:
-                (first-suggestion prompter)
-                next-source)))
-            ((calispel:? (sync-interrupt-channel sync-queue))
-             nil)))
-      ;; No sync-queue if no input was ever set.
-      t)))
+    (block nil
+      (lpara:task-handler-bind ((lpara:task-killed-error (lambda (c)
+                                                           (declare (ignore c))
+                                                           (return nil))))
+        (if (and (all-ready-p prompter :wait-p nil)
+                 (lpara.queue:queue-empty-p (ready-sources prompter)))
+            t
+            (if wait-p
+                (lpara.queue:pop-queue (ready-sources prompter))
+                (lpara.queue:try-pop-queue (ready-sources prompter)
+                                           :timeout 0)))))))
 
 (export-always 'all-ready-p)
-(defun all-ready-p (prompter)
-  "Return non-nil when all PROMPTER sources are ready."
-  (sera:nlet check ((next-source (next-ready-p prompter)))
-    (if (typep next-source 'boolean)
-        next-source
-        (check (next-ready-p prompter)))))
+(defun all-ready-p (prompter &key (wait-p t))
+  "Return non-nil when all PROMPTER sources are ready.
+If WAIT-P, block until all sources are ready."
+  (or (null (kernel prompter))
+      (null (input-reader prompter))
+      (with-kernel prompter
+        (if wait-p
+            ;; TODO: Add `lpara:task-handler-bind' to catch destruction?
+            (and (lpara:force (input-reader prompter))
+                 (lpara:force (source-updater prompter)))
+            (and (lpara:fulfilledp (input-reader prompter))
+                 (lpara:fulfilledp (source-updater prompter)))))))
 
 (export-always 'make)
 (define-function make
```

Comment:
Yet another thing to replace with `nclasses` :P 

And then, source setting can happen in `initialize-instance :after` of `prompter`.

Reply:
How would `nclasses:make-instace*` help here?  Can you give an example?

Reply:
Not really in any way, I guess. Forget it, then!